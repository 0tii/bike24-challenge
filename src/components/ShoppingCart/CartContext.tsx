import { Product } from '@/types/OrderTypes';
import React, { PropsWithChildren, createContext, useState } from 'react';

export interface CartEntry {
  product: Product;
  quantity: number;
}

export interface CartContextType {
  cartItems: CartEntry[];
  addToCart: (item: Product, quantity: number) => boolean;
  removeFromCart: (itemId: string) => void;
  clearCart: () => void;
  modifyItem: (itemId: string, newQuantity: number) => void;
}

export const CartContext = createContext<CartContextType | null>(null);

export const CartProvider = ({ children }: PropsWithChildren) => {
  const [cartItems, setCartItems] = useState<CartEntry[]>([]);

  const addToCart = (item: Product, quantity: number) => {
    // TODO check maxAmount
    setCartItems((prevItems) => [...prevItems, { product: item, quantity: quantity }]);
    return true;
  };

  const removeFromCart = (itemId: string) => {
    setCartItems((prevItems) => prevItems.filter((item) => item.product.id !== itemId));
  };

  const modifyItem = (itemId: string, newQuantity: number) => {
    // TODO To be implemented
  };

  const clearCart = () => {
    setCartItems([]);
  };

  const cartContextValue: CartContextType = {
    cartItems,
    addToCart,
    removeFromCart,
    clearCart,
    modifyItem,
  };

  return <CartContext.Provider value={cartContextValue}>{children}</CartContext.Provider>;
};
