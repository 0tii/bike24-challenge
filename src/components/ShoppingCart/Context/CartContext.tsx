import { Product } from '@/types/OrderTypes';
import React, { PropsWithChildren, createContext, useState } from 'react';
import { constants } from '@/cfg/config';

export interface CartEntry {
  product: Product;
  quantity: number;
}

export interface CartTotals {
  net: number;
  gross: number;
  tax: number;
}

export interface CartContextType {
  cartItems: CartEntry[];
  addToCart: (product: Product, quantity: number) => void;
  removeFromCart: (itemId: string) => void;
  clearCart: () => void;
  modifyItem: (itemId: string, newQuantity: number) => void;
  loadCart: () => void;
  calculateTotals: () => CartTotals;
  orderConfirmation: boolean;
  setOrderConfirmation: (value: boolean) => void;
  cartOpen: boolean;
  setCartOpen: (value: boolean) => void;
}

export const CartContext = createContext<CartContextType | null>(null);

export const CartProvider = ({ children }: PropsWithChildren) => {
  const [cartItems, setCartItems] = useState<CartEntry[]>([]);
  const [orderConfirmation, setOrderConfirmation] = useState(false);
  const [cartOpen, setCartOpen] = useState(false);

  /**
   * @throws Cart is full, Quantity exceeds allowed limit
   */
  const addToCart = (product: Product, quantity: number) => {
    if (cartItems.length >= constants.MAX_CART_ITEMS) throw 'The cart is full.';

    const itemIndex = cartItems.findIndex((item) => item.product.id === product.id);

    if (itemIndex === -1) {
      setCartItems((prevItems) => [...prevItems, { product: product, quantity: quantity }]);
    } else {
      const cartCopy = [...cartItems];
      if (cartCopy[itemIndex].quantity + quantity > product.maxAmount)
        throw 'Total quantity for this product exceeds allowed limit';

      cartCopy[itemIndex].quantity += quantity;
      setCartItems(cartCopy);
    }

    saveCart();
  };

  const calculateTotals = () => {
    let totalNet = 0,
      totalTax = 0,
      totalGross = 0;

    cartItems.forEach((i) => {
      let currentTax = ((i.product.price * i.product.taxRate) / 100) * i.quantity;
      totalNet += i.product.price * i.quantity;
      totalTax += currentTax;
      totalGross += i.product.price * i.quantity + currentTax;
    });

    return {
      net: totalNet,
      gross: totalGross,
      tax: totalTax,
    };
  };

  const removeFromCart = (itemId: string) => {
    setCartItems((prevItems) => prevItems.filter((item) => item.product.id !== itemId));
    saveCart();
  };

  /**
   * @throws Item could not be found
   */
  const modifyItem = (itemId: string, newQuantity: number) => {
    const cartCopy = [...cartItems];
    const targetIndex = cartCopy.findIndex((item) => item.product.id === itemId);

    if (targetIndex === -1) throw 'The item could not be found.';

    cartCopy[targetIndex].quantity = newQuantity;
    setCartItems(cartCopy);
    saveCart();
  };

  const clearCart = () => {
    setCartItems([]);
    saveCart(true);
  };

  const saveCart = (clear = false) => {
    if (clear) return localStorage.removeItem('cart');
    localStorage.setItem('cart', Buffer.from(JSON.stringify(cartItems), 'utf8').toString('base64'));
  };

  /**
   * @throws No cart available, Could not deserialize cart items
   */
  const loadCart = () => {
    const b64 = localStorage.getItem('cart');
    if (!b64) throw 'No cart available.';
    try {
      const cart = JSON.parse(Buffer.from(b64, 'base64').toString('utf8'));
      setCartItems(cart);
    } catch (err) {
      throw 'Could not deserialize the cart items.';
    }
  };

  const cartContextValue: CartContextType = {
    cartItems,
    addToCart,
    removeFromCart,
    clearCart,
    modifyItem,
    loadCart,
    calculateTotals,
    orderConfirmation,
    setOrderConfirmation,
    cartOpen,
    setCartOpen,
  };

  return <CartContext.Provider value={cartContextValue}>{children}</CartContext.Provider>;
};
