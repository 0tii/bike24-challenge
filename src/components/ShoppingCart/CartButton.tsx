import ShoppingCartIcon from '@/svg/ShoppingCartIcon';
import { CartContext } from './CartContext';
import { HTMLProps, useContext } from 'react';

interface CartButtonProps extends HTMLProps<HTMLDivElement> {}

export const CartButton = ({ onClick, className, ...props }: CartButtonProps) => {
  const shoppingCart = useContext(CartContext);

  return (
    <div {...props} className={`${className} relative flex flex-col justify-center max-w-[80px]`}>
      <button
        className="bg-blue-800 hover:bg-blue-700 rounded-md px-8 py-3 w-full"
        onClick={() => shoppingCart?.setCartOpen(true)}
        aria-label="open shopping cart button"
      >
        <ShoppingCartIcon className="fill-white scale-150 w-full" />
      </button>

      <div
        className="absolute flex items-center justify-center rounded-full bg-orange-600 text-white text-bold text-xs top-[-8px] right-[-5px] h-6 w-6"
        aria-label="amount of products in cart"
      >
        {shoppingCart?.cartItems.length}
      </div>

      {shoppingCart?.cartItems.length !== undefined && shoppingCart?.cartItems.length > 0 && (
        <div
          className="flex items-center justify-center rounded-full bg-sky-600 text-white text-bold text-xs p-[2px] mt-1"
          aria-label="cart total sum"
        >
          {shoppingCart?.calculateTotals().gross.toFixed(2)}â‚¬
        </div>
      )}
    </div>
  );
};
