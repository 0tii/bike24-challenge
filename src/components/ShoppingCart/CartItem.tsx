import { Product } from '@/types/OrderTypes';
import { useContext } from 'react';
import { CartContext } from './CartContext';
import TrashBinIcon from '@/svg/TrashBinIcon';

export interface CartItemProps {
  product: Product;
  quantity: number;
}

export const CartItem = ({ product, quantity }: CartItemProps) => {
  const shoppingCart = useContext(CartContext);

  return (
    <div className="grid grid-cols-[40%_auto] text-sm sm:text-base bg-slate-300 rounded-md p-3 mt-2">
      <span>{product.productName}</span>
      <div className="flex flex-row flex-1 justify-between">
        <span>{quantity}</span>
        <span>{product.price}€</span>
        <span>
          <b>{(quantity * product.price).toFixed(2)}€</b>
        </span>
        <button
          className="bg-red-500 hover:bg-red-600 rounded-md p-2"
          onClick={() => shoppingCart?.removeFromCart(product.id)}
        >
          <TrashBinIcon className=" fill-white w-5 h-5 stroke-2" />
        </button>
      </div>
    </div>
  );
};

export default CartItem;
