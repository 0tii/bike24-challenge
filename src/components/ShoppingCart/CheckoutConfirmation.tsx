import ShopSuccessIcon from '@/svg/ShopSuccessIcon';
import { useContext } from 'react';
import { CartContext } from './CartContext';

interface CheckoutConfirmationProps {
  onConfirm: () => void;
}

export const CheckoutConfirmation = ({ onConfirm }: CheckoutConfirmationProps) => {
  const cart = useContext(CartContext);
  return (
    <>
      <div className="absolute z-[19] w-screen h-screen top-0 left-0 bg-black/60" />
      <div className="z-20 absolute bg-emerald-300 border-green-100 border-[2px] w-full sm:w-[40%] sm:min-w-[200px] sm:max-w-[400px] h-[40%] top-1/2 left-1/2 transform -translate-x-1/2 -translate-y-1/2 rounded-md">
        <div className="flex flex-col w-full h-full justify-between items-center py-8">
          <h1 className=" font-bold text-xl">Thank you for your purchase!</h1>
          <ShopSuccessIcon className="w-24 h-24 fill-emerald-600" />
          <span className="font-semibold text-lg">
            <i>Continue shopping</i>
          </span>
          <button
            className="p-3 bg-gray-200 rounded-md hover:bg-gray-300"
            onClick={() => {
              onConfirm();
              cart?.clearCart();
            }}
          >
            Continue
          </button>
        </div>
      </div>
    </>
  );
};

export default CheckoutConfirmation;
