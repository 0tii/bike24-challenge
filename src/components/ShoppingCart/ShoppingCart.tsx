import { Product } from '@/types/OrderTypes';
import { useContext, useEffect, useState } from 'react';
import { CartContext, CartTotals } from './CartContext';
import { constants } from '@/cfg/config';
import CartItemList from './CartList/CartItemList';
import CartListTotals from './CartList/CartListTotals';
import { toast } from 'react-toastify';
import ClearShoppingCartIcon from '@/svg/ClearShoppingCartIcon';

export interface ShoppingCartItem {
  product: Product;
  quantity: number;
}

interface ShoppingCartProps {
  // ...
}

const ShoppingCart = ({}: ShoppingCartProps) => {
  const shoppingCart = useContext(CartContext);

  const onClear = () => {
    if (!shoppingCart || shoppingCart.cartItems.length === 0) {
      toast('The shopping cart is already empty.', {
        type: 'error',
        autoClose: 1400,
      });
      return;
    }

    shoppingCart?.clearCart();
    toast('Shopping cart cleared.', {
      type: 'success',
      autoClose: 1400,
    });
  };

  return (
    <>
      <h2 className="my-3 font-bold text-4xl">Your orders</h2>
      <CartItemList shoppingCart={shoppingCart} />
      {shoppingCart?.cartItems.length === 0 && (
        <div className="flex flex-col items-center my-8 text-gray-300 font-bold text-[2rem]">
          <span>Add products first</span>
          <ClearShoppingCartIcon className="fill-gray-300 h-24 w-24" />
        </div>
      )}
      <CartListTotals shoppingCart={shoppingCart} />

      <div className="flex flex-row justify-between items-end">
        <button
          aria-label="clear cart button"
          className="bg-gray-500 hover:bg-gray-600 text-white rounded-md py-3 px-5"
          onClick={onClear}
        >
          Clear
        </button>
        <div className=" w-full max-w-[150px]">
          <progress
            aria-label="product count progress bar"
            className="max-w-[150px] [&::-webkit-progress-bar]:rounded-md [&::-webkit-progress-value]:rounded-md [&::-webkit-progress-bar]:bg-slate-300 [&::-webkit-progress-value]:bg-blue-400 [&::-moz-progress-bar]:bg-blue-400"
            value={shoppingCart?.cartItems.length}
            max={constants.MAX_CART_ITEMS}
          />
          <button
            aria-label="checkout button"
            className="rounded-md bg-blue-500 text-white p-4 hover:bg-blue-600 disabled:bg-gray-400 font-bold w-full"
          >
            Checkout
          </button>
        </div>
      </div>
    </>
  );
};

export default ShoppingCart;
