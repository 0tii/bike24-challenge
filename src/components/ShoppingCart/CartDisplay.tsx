import { useState } from 'react';
import ShoppingCart from './ShoppingCart';
import { CartButton } from './CartButton';

export const CartDisplay = () => {
  const [expanded, setExpanded] = useState(false);

  return (
    <>
      <CartButton onClick={() => setExpanded((exp) => !exp)} aria-label="shopping cart button" />
      {expanded && (
        <>
          <div className="absolute h-screen w-screen top-0 left-0 bg-[rgba(61,61,61,0.34)]" />

          <div
            className="slidein z-10 top-0 right-0 absolute md:w-[60%] md:max-w-[650px] w-full min-w-[300px] h-full bg-white overflow-y-auto overflow-x-hidden shadow-2xl"
            aria-label="shopping cart screen"
          >
            <div className="flex flex-col">
              <div
                className="bg-blue-800 w-full flex justify-end items-center"
                aria-label="cart header"
              >
                <button
                  className="text-white text-xl mr-5 py-3"
                  aria-label="close cart button"
                  onClick={() => setExpanded((exp) => !exp)}
                >
                  x
                </button>
              </div>
              <div className="w-full px-3">
                <ShoppingCart />
              </div>
            </div>
          </div>
        </>
      )}
    </>
  );
};

export default CartDisplay;
