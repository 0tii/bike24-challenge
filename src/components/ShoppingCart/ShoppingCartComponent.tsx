import { Product } from '@/types/OrderTypes';
import { useCallback, useContext } from 'react';
import { CartContext } from './CartContext';
import CartItem from './CartItem';

export interface ShoppingCartItem {
  product: Product;
  quantity: number;
}

export interface ShoppingCartComponentProps {
  // ...
}

const ShoppingCartComponent = ({}: ShoppingCartComponentProps) => {
  const shoppingCart = useContext(CartContext);

  return (
    <>
      {shoppingCart?.cartItems.map((item) => (
        <CartItem key={item.product.id} product={item.product} quantity={item.quantity} />
      ))}
    </>
  );
};

export default ShoppingCartComponent;
