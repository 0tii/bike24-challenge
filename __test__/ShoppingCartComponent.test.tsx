import { render, fireEvent, within } from '@testing-library/react';
import '@testing-library/jest-dom';
import { CartContext, CartContextType } from '@/components/ShoppingCart/CartContext';
import ShoppingCartComponent from '@/components/ShoppingCart/ShoppingCartComponent';

describe('Shopping Cart Component', () => {
  it('renders', () => {
    const container = render(
      <CartContext.Provider value={mockCartContext}>
        <ShoppingCartComponent />
      </CartContext.Provider>
    );

    expect(container).toMatchSnapshot();
  });

  it('renders all elements in the cart as table rows', () => {
    const { getByRole } = render(
      <CartContext.Provider value={mockCartContext}>
        <ShoppingCartComponent />
      </CartContext.Provider>
    );

    const table = getByRole('table');
    const rows = within(table).getAllByRole('row');

    expect(rows).toHaveLength(mockCartContext.cartItems.length + 1);
  });
});

const mockProducts = [
  {
    id: '8e3973e3-e43e-4370-bfcc-3b84d72bb77d',
    productName: 'Syrup - Monin, Irish Cream',
    maxAmount: 3,
    taxRate: 7,
    price: 7.76,
  },
  {
    id: 'e713cc29-9074-49f1-9e9c-d7b340fac027',
    productName: 'Pie Shells 10',
    maxAmount: 94,
    taxRate: 7,
    price: 9.45,
  },
  {
    id: '4f6dff49-9731-4bfa-9e67-854238328a80',
    productName: 'Wine - Trimbach Pinot Blanc',
    maxAmount: 17,
    taxRate: 7,
    price: 14.31,
  },
];

const mockCartContext: CartContextType = {
  cartItems: [
    { product: mockProducts[0], quantity: 2 },
    { product: mockProducts[1], quantity: 50 },
  ],
  addToCart: jest.fn(),
  removeFromCart: jest.fn(),
  clearCart: jest.fn(),
  modifyItem: jest.fn(),
  loadCart: jest.fn(),
  calculateTotals: jest.fn(),
};
