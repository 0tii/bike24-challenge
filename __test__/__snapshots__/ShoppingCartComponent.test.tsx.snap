// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Shopping Cart Component renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <h2
        class="my-3 font-bold text-4xl"
      >
        Your orders
      </h2>
      <table
        class="w-full sm:text-base text-sm overflow-x-auto"
      >
        <thead
          class="sm:text-base text-xs"
        >
          <tr>
            <th
              class="text-start"
            >
              Product
            </th>
            <th
              class="text-start max-w-[35px] sm:max-w-auto"
            >
              <div
                class="whitespace-nowrap truncate w-full"
              >
                Quantity
              </div>
            </th>
            <th
              class="text-start"
            >
              Price
            </th>
            <th
              class="text-start"
            >
              Total
            </th>
            <th
              class="text-start"
            />
          </tr>
        </thead>
        <tbody>
          <tr>
            <td
              class="sm:py-2 py-1"
            >
              Syrup - Monin, Irish Cream
            </td>
            <td
              class="sm:py-2 py-1"
            >
              2
            </td>
            <td
              class="sm:py-2 py-1"
            >
              7.76
              €
            </td>
            <td
              class="sm:py-2 py-1"
            >
              15.52
              €
            </td>
            <td
              class="sm:py-2 py-1"
            >
              <div
                class="flex justify-end"
              >
                <button
                  aria-label="delete button"
                  class="bg-red-500 hover:bg-red-600 rounded-md sm:p-2 p-1"
                >
                  <svg
                    class="fill-white sm:w-5 sm:h-5 w-4 h-4"
                    viewBox="0 -960 960 960"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M261-120q-24.75 0-42.375-17.625T201-180v-570h-41v-60h188v-30h264v30h188v60h-41v570q0 24-18 42t-42 18H261Zm438-630H261v570h438v-570ZM367-266h60v-399h-60v399Zm166 0h60v-399h-60v399ZM261-750v570-570Z"
                    />
                  </svg>
                </button>
              </div>
            </td>
          </tr>
          <tr>
            <td
              class="sm:py-2 py-1"
            >
              Pie Shells 10
            </td>
            <td
              class="sm:py-2 py-1"
            >
              50
            </td>
            <td
              class="sm:py-2 py-1"
            >
              9.45
              €
            </td>
            <td
              class="sm:py-2 py-1"
            >
              472.50
              €
            </td>
            <td
              class="sm:py-2 py-1"
            >
              <div
                class="flex justify-end"
              >
                <button
                  aria-label="delete button"
                  class="bg-red-500 hover:bg-red-600 rounded-md sm:p-2 p-1"
                >
                  <svg
                    class="fill-white sm:w-5 sm:h-5 w-4 h-4"
                    viewBox="0 -960 960 960"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M261-120q-24.75 0-42.375-17.625T201-180v-570h-41v-60h188v-30h264v30h188v60h-41v570q0 24-18 42t-42 18H261Zm438-630H261v570h438v-570ZM367-266h60v-399h-60v399Zm166 0h60v-399h-60v399ZM261-750v570-570Z"
                    />
                  </svg>
                </button>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
      <div
        class="flex flex-row justify-end text-sm sm:text-base bg-gray-100 rounded-sm p-3 my-4"
      >
        <div>
          <div
            class="grid grid-cols-2 text-gray-600"
          >
            <span>
              Subtotal:
            </span>
            <span
              class="ml-3 text-end"
            >
              €
            </span>
          </div>
          <div
            class="grid grid-cols-2 text-gray-600"
          >
            <span>
              Tax:
            </span>
            <span
              class="ml-3 text-end"
            >
              €
            </span>
          </div>
          <div
            class="grid grid-cols-2 font-bold"
          >
            <span>
              Gross Total:
            </span>
            <span
              class="ml-3 text-end"
            >
              €
            </span>
          </div>
        </div>
      </div>
      <div
        class="flex flex-row justify-between items-end"
      >
        <button
          aria-label="clear cart button"
          class="bg-gray-500 hover:bg-gray-600 text-white rounded-md py-3 px-5"
        >
          Clear
        </button>
        <div
          class=" w-full max-w-[150px]"
        >
          <progress
            aria-label="product count progress bar"
            class="max-w-[150px] [&::-webkit-progress-bar]:rounded-md [&::-webkit-progress-value]:rounded-md [&::-webkit-progress-bar]:bg-slate-300 [&::-webkit-progress-value]:bg-blue-400 [&::-moz-progress-bar]:bg-blue-400"
            max="10"
            value="2"
          />
          <button
            aria-label="checkout button"
            class="rounded-md bg-blue-500 text-white p-4 hover:bg-blue-600 disabled:bg-gray-400 font-bold w-full"
          >
            Checkout
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <h2
      class="my-3 font-bold text-4xl"
    >
      Your orders
    </h2>
    <table
      class="w-full sm:text-base text-sm overflow-x-auto"
    >
      <thead
        class="sm:text-base text-xs"
      >
        <tr>
          <th
            class="text-start"
          >
            Product
          </th>
          <th
            class="text-start max-w-[35px] sm:max-w-auto"
          >
            <div
              class="whitespace-nowrap truncate w-full"
            >
              Quantity
            </div>
          </th>
          <th
            class="text-start"
          >
            Price
          </th>
          <th
            class="text-start"
          >
            Total
          </th>
          <th
            class="text-start"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="sm:py-2 py-1"
          >
            Syrup - Monin, Irish Cream
          </td>
          <td
            class="sm:py-2 py-1"
          >
            2
          </td>
          <td
            class="sm:py-2 py-1"
          >
            7.76
            €
          </td>
          <td
            class="sm:py-2 py-1"
          >
            15.52
            €
          </td>
          <td
            class="sm:py-2 py-1"
          >
            <div
              class="flex justify-end"
            >
              <button
                aria-label="delete button"
                class="bg-red-500 hover:bg-red-600 rounded-md sm:p-2 p-1"
              >
                <svg
                  class="fill-white sm:w-5 sm:h-5 w-4 h-4"
                  viewBox="0 -960 960 960"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M261-120q-24.75 0-42.375-17.625T201-180v-570h-41v-60h188v-30h264v30h188v60h-41v570q0 24-18 42t-42 18H261Zm438-630H261v570h438v-570ZM367-266h60v-399h-60v399Zm166 0h60v-399h-60v399ZM261-750v570-570Z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
        <tr>
          <td
            class="sm:py-2 py-1"
          >
            Pie Shells 10
          </td>
          <td
            class="sm:py-2 py-1"
          >
            50
          </td>
          <td
            class="sm:py-2 py-1"
          >
            9.45
            €
          </td>
          <td
            class="sm:py-2 py-1"
          >
            472.50
            €
          </td>
          <td
            class="sm:py-2 py-1"
          >
            <div
              class="flex justify-end"
            >
              <button
                aria-label="delete button"
                class="bg-red-500 hover:bg-red-600 rounded-md sm:p-2 p-1"
              >
                <svg
                  class="fill-white sm:w-5 sm:h-5 w-4 h-4"
                  viewBox="0 -960 960 960"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M261-120q-24.75 0-42.375-17.625T201-180v-570h-41v-60h188v-30h264v30h188v60h-41v570q0 24-18 42t-42 18H261Zm438-630H261v570h438v-570ZM367-266h60v-399h-60v399Zm166 0h60v-399h-60v399ZM261-750v570-570Z"
                  />
                </svg>
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="flex flex-row justify-end text-sm sm:text-base bg-gray-100 rounded-sm p-3 my-4"
    >
      <div>
        <div
          class="grid grid-cols-2 text-gray-600"
        >
          <span>
            Subtotal:
          </span>
          <span
            class="ml-3 text-end"
          >
            €
          </span>
        </div>
        <div
          class="grid grid-cols-2 text-gray-600"
        >
          <span>
            Tax:
          </span>
          <span
            class="ml-3 text-end"
          >
            €
          </span>
        </div>
        <div
          class="grid grid-cols-2 font-bold"
        >
          <span>
            Gross Total:
          </span>
          <span
            class="ml-3 text-end"
          >
            €
          </span>
        </div>
      </div>
    </div>
    <div
      class="flex flex-row justify-between items-end"
    >
      <button
        aria-label="clear cart button"
        class="bg-gray-500 hover:bg-gray-600 text-white rounded-md py-3 px-5"
      >
        Clear
      </button>
      <div
        class=" w-full max-w-[150px]"
      >
        <progress
          aria-label="product count progress bar"
          class="max-w-[150px] [&::-webkit-progress-bar]:rounded-md [&::-webkit-progress-value]:rounded-md [&::-webkit-progress-bar]:bg-slate-300 [&::-webkit-progress-value]:bg-blue-400 [&::-moz-progress-bar]:bg-blue-400"
          max="10"
          value="2"
        />
        <button
          aria-label="checkout button"
          class="rounded-md bg-blue-500 text-white p-4 hover:bg-blue-600 disabled:bg-gray-400 font-bold w-full"
        >
          Checkout
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
